File: src/hash_table/two_sum.py
from typing import List


class TwoSum:
    """
    Problem Link: https://leetcode.com/problems/two-sum/
    Complexity: Easy
    """
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        available_numbers = {num: idx for idx, num in enumerate(nums)}

        for idx, num in enumerate(nums):
            remaining_number = target - num

            if remaining_number in available_numbers and available_numbers[remaining_number] != idx:
                remaining_idx = available_numbers[remaining_number]

                return idx, remaining_idx

        return ()

File: src/hash_table/two_sum/two_sum.go
package two_sum

func twoSum(nums []int, target int) []int {
	numMap := make(map[int]int, len(nums)) // number -> index

	for idx, n := range nums {
		numMap[n] = idx
	}

	for i, n := range nums {
		remainder := target - n

		if j, ok := numMap[remainder]; ok && j != i {
			return []int{i, j}
		}
	}

	return []int{}
}
